### PR TITLE
Identify compatibility check v2

### LINKED ISSUE #1769: Design proposal for `/itchysats/identify` protocol
The proposed design is loosely based on https://github.com/libp2p/specs/tree/master/identify.

Core design choices:

- Only support _pulling_ identify information: Identify information is primarily useful for the peer that receives it. We don't want to push information to peers unless we know they are interested in it.
- Identify information should include:
	- agent-version
	- list of supported protocols
- Use JSON for encoding messages. We can switch a binary format in a different version.

The agent-version is useful for monitoring purposes.

Retrieving the list of supported protocol will allow us to make a runtime decision on which protocols to use to communicate with a peer. For example, a peer can signal that they want to have offers pushed (see https://github.com/itchysats/itchysats/issues/1768) by advertising the `/itchysats/offer` protocol.

Currently, `libp2p-xtra` has a static configuration of which protocols are supported by an `Endpoint`. However, one can imagine making this list dynamic and thus, advertising different protocols to different peers. This will allow for a form of A/B testing where we can offer a specific protocol to a few select peers and/or to a subset (i.e. 10%).

For example, a maker could advertise that it supports a new rollover protocol to select peers, thus allowing for a way of safely testing things in production without rolling them out to all takers at once.

## Open questions

- Use protocol specified in https://github.com/libp2p/specs/tree/master/identify instead? It will require the use of protobuf for the message encoding.

### LINKED ISSUE #2307: Provide feedback about protocol between maker/taker incompatibility based on `identify` protocol
Follow up of #2161 that adds the "compatibility" check between taker and maker.
This goes beyond the feature set we had in the old legacy stack, because we are able to do this in a better way now.

`identify` includes the list of protocols that another peer supports, i.e. the protocols that one can open a substream for with the peer.
Given this, we can actually write *smarter* compatibility code once a taker has information about he maker.

---

**Example:**

Let's say we removed a legacy version of the `itchysats/collaborative_settlement` protocol, the maker now only supports `itchysats/collaborative_settlement/2.0.0`, but not `itchysats/collaborative_settlement/1.0.0`.

A taker that only supports `itchysats/collaborative_settlement/1.0.0` will notice that there is no compatible collab settlement protocol supported by the maker - but that does not mean we have to drop the connection!
Rollovers are not affected, so the taker does not have to upgrate *immediately* - but has to in order to be able to collab close the position.

---

This will potentially allow us to be more graceful with protocol upgrades, if we decide to implement logic (and UI changes for it).
These changes can potentially become complex, because we actually have dependencies between protocols that might not be obvious, e.g. `itchysats/offers` is needed for various other protocols, because without an up-to-date offer we cannot do e.g. rollovers.

---

As a simplified first step I think it would be useful to just display incompatibility in the taker UI. I would not close the connection, but accept that certain protocols may fail when trying to open a substream if the protocol is not supported on the other side. 
From the user's perspective this will look similar to the legacy behavior (i.e. `your daemon is not compatible, update (...)!`) but it is actually more resilient because we don't have to close the connection if we have incompatible protocols. The incompatible protocol can just fail (while the others still succeed!).

Out of scope:

Over time we can decide to make this smarter and e.g. disable parts of the UI based on the protocol information we have.
Over time we *could* define scenarios that are "breaking" and actually drop a connection to such a peer.

### PR DESCRIPTION
**Pull request by luckysori**
Fixes #1769.
Fixes #2307.

Hopefully the same stuff as in #2383, but with the listener protocol compatibility modelled differently.

I also addressed a couple of my own comments from the original PR.

### CODE REVIEW

File: daemon/src/identify/dialer.rs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+pub(crate) struct GetPeerInfo(pub PeerId);
+
+pub(crate) struct IdentifyMsgReceived {
+    peer_id: PeerId,
+    identify_msg: protocol::IdentifyMsg,
+}
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle(&mut self, msg: GetPeerInfo) -> Option<PeerInfo> {
```
Comment (da-kami) [inline daemon/src/identify/dialer.rs]: We should be able to remove this - if I forgot to remove it on #2383 then that was by accident.
After the channel was introduced this is not used.

**Review by da-kami — COMMENTED** [review]
I am happy to see a refactoring, but I am confused by this PR.

Are we expecting to close #2383 in favor of this? - it might be cleaner to create a PR on top of #2383 that showcases the changes. Like this it is super hard to know what changed.
Could you maybe change it so it is a PR on top of #2383 rather than replacing it? Otherwise I'm happy to talk this through together tomorrow.

**Comment by luckysori** [comment]
This work has been force-pushed onto https://github.com/itchysats/itchysats/pull/2383 so that it can be reviewed.

**Comment by klochowicz** [comment]
@luckysori it's super hard to compare the 2 approaches when it got force-pushed to the other PR - any chance it could have been added as a PR against @da-kami 's PR?

**Pull request closed without merging**